src/peer_manager.rs

```diff
@@ -296,17 +311,184 @@ impl PeerManager {
+    }
+
+    pub fn verify_candidate(&mut self,
+                            node_name: XorName,
+                            proof: Vec<u8>,
+                            _leading_zero_bytes: u32,
+                            _difficulty: u32)
+                            -> Option<bool> {
+        if let Some((ref name, ref now, ref seed)) = self.node_candidate {
+            if *name == node_name &&
+               now.elapsed() < Duration::from_secs(RESOURCE_PROOF_EVALUATE_TIMEOUT_SECS) {
+                return Some(*seed == proof);
+            }
+        }
+        None
+    }
+
+    pub fn handle_node_approval_vote(&mut self,
+                                     candidate_name: XorName,
+                                     validity: bool)
+                                     -> (bool, Option<(PublicId, PeerId)>) {
+        if let Some((name, _, _)) = self.node_candidate {
+            if name == candidate_name {
+                let result = if validity {
+                    let (pub_id, peer_id) = if let Some(peer) = self.peer_map
+                        .get_by_name(&candidate_name) {
+                        if let Some(peer_id) = peer.peer_id().clone() {
+                            (*peer.pub_id(), *peer_id)
+                        } else {
+                            return (false, None);
+                        }
+                    } else {
+                        return (false, None);
+                    };
+                    Some((pub_id, peer_id))
+                } else {
+                    None
+                };
+                self.node_candidate = None;
+                return (true, result);
+            }
+        }
+        (false, None)
+    }
+
+    pub fn handle_node_approval(&mut self) -> Vec<(PublicId, PeerId)> {
+        let mut result = vec![];
+        for peer_name in &self.peer_candidates.clone() {
+            if let Some(peer) = self.peer_map.get_by_name(peer_name) {
+                if let Some(peer_id) = peer.peer_id().clone() {
+                    result.push((*peer.pub_id(), *peer_id));
+                }
+            }
+        }
+        self.peer_candidates.clear();
+        result
+    }
+
+    pub fn is_node_candidate(&mut self,
+                             pub_id: &PublicId,
+                             peer_id: &PeerId)
+                             -> Result<Option<(bool, Vec<u8>)>, RoutingTableError> {
+        if let Some((ref name, _, ref seed)) = self.node_candidate {
+            if name == pub_id.name() {
+                let tunnel = match self.peer_map.remove(peer_id).map(|peer| peer.state) {
+                    Some(PeerState::SearchingForTunnel) |
+                    Some(PeerState::AwaitingNodeIdentify(true)) => true,
+                    Some(PeerState::Routing(_tunnel)) => {
+                        error!("NodeCandidate {:?} already in state Routing.", peer_id);
+                        return Err(RoutingTableError::AlreadyExists);
```

Comment:
If the function returns here the peer gets removed from the `peer_map` without replacement. I don't think this was intended?